File: src/tools/finance/backtest/baselines.ts
/**
 * R5 Idea #3 — Naive baseline guards for forecast quality CI.
 *
 * Source: docs/forecast-improvement-ideas-round5-2026-04-29.md (Idea #3),
 * arXiv:2502.09079 (Puoti et al. 2025) — "Crypto unpredictability".
 *
 * Computes simple model-free baselines that any non-trivial forecast head
 * must beat:
 *   - Coin-flip:    always P(up)=0.5, recommends HOLD
 *   - Last-period:  P(up)=1 if previous step was up, recommends BUY/SELL
 *
 * Returned metrics use the same {@link BacktestStep} shape so they can be
 * compared head-to-head with `directionalAccuracy`/`brierScore` from
 * `metrics.ts`.
 *
 * The CI gate (used in `markov-backtest.integration.test.ts`) is:
 *   "for every horizon h ≥ 7d, the deployed arm's `directionalAccuracy`
 *   must be ≥ baseline.lastPeriod.directionalAccuracy − 0.02".  We allow
 *   a 2pp slack to absorb sampling noise on small fixtures.
 */
import { brierScore, directionalAccuracy, type BacktestStep } from './metrics.js';

export interface BaselineMetricBlock {
  n: number;
  brierScore: number;
  directionalAccuracy: number;
}

export interface NaiveBaselineReport {
  /** "Always P=0.5, recommend HOLD" — the strict random null. */
  coinFlip: BaselineMetricBlock;
  /** "Predict last realized direction continues" — momentum-1 baseline. */
  lastPeriod: BaselineMetricBlock;
}

/**
 * Build a synthetic step that mirrors `actual` but with the synthetic
 * baseline prediction, so we can reuse the existing metric implementations.
 */
function synth(
  step: BacktestStep,
  predictedProb: number,
  recommendation: BacktestStep['recommendation'],
): BacktestStep {
  return {
    ...step,
    predictedProb,
    rawPredictedProb: predictedProb,
    recommendation,
    confidence: 0.5,
    gofPasses: null,
  };
}

export function computeCoinFlipBaseline(steps: BacktestStep[]): BaselineMetricBlock {
  if (steps.length === 0) return { n: 0, brierScore: 1, directionalAccuracy: 0 };
  const synthetic = steps.map(s => synth(s, 0.5, 'HOLD'));
  return {
    n: steps.length,
    brierScore: brierScore(synthetic),
    directionalAccuracy: directionalAccuracy(synthetic),
  };
}

/**
 * "Last period" baseline: predicts the direction of the *previous*
 * realized return continues. For the first step (no prior), it falls
 * back to the coin-flip prior (P=0.5, HOLD).
 *
 * For probability assignment we use a confident-but-not-degenerate value
 * (0.65 / 0.35) — picking 1.0 / 0.0 would hand the baseline an unfair
 * Brier score (0 when right, 1 when wrong, dominated by the 50%
 * directional accuracy seen on random walks).
 */
export function computeLastPeriodBaseline(steps: BacktestStep[]): BaselineMetricBlock {
  if (steps.length === 0) return { n: 0, brierScore: 1, directionalAccuracy: 0 };
  let prevReturn = 0;
  const synthetic: BacktestStep[] = steps.map((step, i) => {
    if (i === 0) {
      prevReturn = step.actualReturn;
      return synth(step, 0.5, 'HOLD');
    }
    const predUp = prevReturn > 0;
    const predictedProb = predUp ? 0.65 : 0.35;
    const recommendation = predUp ? 'BUY' : 'SELL';
    prevReturn = step.actualReturn;
    return synth(step, predictedProb, recommendation);
  });
  return {
    n: steps.length,
    brierScore: brierScore(synthetic),
    directionalAccuracy: directionalAccuracy(synthetic),
  };
}

export function computeNaiveBaselines(steps: BacktestStep[]): NaiveBaselineReport {
  return {
    coinFlip: computeCoinFlipBaseline(steps),
    lastPeriod: computeLastPeriodBaseline(steps),
  };
}

/**
 * Strict gate: deployed arm must have directional accuracy at least as
 * good as the better of the two naive baselines, minus a slack.
 *
 * @param armSteps      — steps from the candidate arm
 * @param slack         — pp slack absorbed (default 0.02)
 * @returns             — {passes, armDirAcc, baselineDirAcc, gap}
 */
export function naiveBaselineGuard(
  armSteps: BacktestStep[],
  slack = 0.02,
): { passes: boolean; armDirAcc: number; baselineDirAcc: number; gap: number } {
  const armDirAcc = directionalAccuracy(armSteps);
  const baselines = computeNaiveBaselines(armSteps);
  const baselineDirAcc = Math.max(
    baselines.coinFlip.directionalAccuracy,
    baselines.lastPeriod.directionalAccuracy,
  );
  const gap = armDirAcc - baselineDirAcc;
  return { passes: gap >= -slack, armDirAcc, baselineDirAcc, gap };
}

File: research/backtest/baselines.py
"""R5 Idea #3 — Naive baseline guards for forecast quality.

Python mirror of src/tools/finance/backtest/baselines.ts.

Source: docs/forecast-improvement-ideas-round5-2026-04-29.md (Idea #3),
arXiv:2502.09079 Puoti et al. 2025.

Computes simple model-free baselines that any non-trivial forecast head
must beat:
  - Coin-flip:   always P(up)=0.5
  - Last-period: P(up)=0.65 if previous step was up, else 0.35

The gate: deployed arm directional accuracy must be >= better baseline
accuracy minus `slack` (default 0.02).
"""

from __future__ import annotations
from dataclasses import dataclass
from research.backtest.walk_forward import BacktestStep


@dataclass(frozen=True)
class BaselineMetricBlock:
    """Brier score + directional accuracy for a synthetic baseline."""

    n: int
    brier_score: float
    directional_accuracy: float


@dataclass(frozen=True)
class NaiveBaselineReport:
    """Coin-flip and last-period baseline metrics."""

    coin_flip: BaselineMetricBlock
    last_period: BaselineMetricBlock


def _synth(step: BacktestStep, predicted_prob: float) -> BacktestStep:
    """Return a copy of `step` with an overridden predicted_prob and
    direction_correct recomputed from the new probability."""
    pred_up = predicted_prob > 0.5
    direction_correct = (pred_up and step.realised_return > 0) or (
        not pred_up and step.realised_return <= 0
    )
    return BacktestStep(
        start_idx=step.start_idx,
        predicted_prob=predicted_prob,
        predicted_return=step.predicted_return,
        ci_lower=step.ci_lower,
        ci_upper=step.ci_upper,
        realised_return=step.realised_return,
        realised_price=step.realised_price,
        direction_correct=direction_correct,
        in_ci=step.in_ci,
    )


def compute_coin_flip_baseline(steps: list[BacktestStep]) -> BaselineMetricBlock:
    """Baseline: always predict P(up)=0.5 (strict random null)."""
    if not steps:
        return BaselineMetricBlock(n=0, brier_score=1.0, directional_accuracy=0.0)

    synthetic = [_synth(s, 0.5) for s in steps]
    outcomes = [1.0 if s.realised_return > 0 else 0.0 for s in synthetic]

    bs = sum((s.predicted_prob - o) ** 2 for s, o in zip(synthetic, outcomes)) / len(synthetic)
    da = sum(1 for s in synthetic if s.direction_correct) / len(synthetic)

    return BaselineMetricBlock(n=len(steps), brier_score=bs, directional_accuracy=da)


def compute_last_period_baseline(steps: list[BacktestStep]) -> BaselineMetricBlock:
    """Baseline: predict last realized direction continues.

    For step 0 (no prior): falls back to coin-flip (P=0.5).
    Probability assignment: 0.65 if last-up, 0.35 if last-down (avoids
    degenerate 0/1 Brier extremes on random walks).
    """
    if not steps:
        return BaselineMetricBlock(n=0, brier_score=1.0, directional_accuracy=0.0)

    synthetic: list[BacktestStep] = []
    prev_return = 0.0

    for i, step in enumerate(steps):
        if i == 0:
            synth_step = _synth(step, 0.5)
        else:
            prob = 0.65 if prev_return > 0 else 0.35
            synth_step = _synth(step, prob)
        prev_return = step.realised_return
        synthetic.append(synth_step)

    outcomes = [1.0 if s.realised_return > 0 else 0.0 for s in synthetic]

    bs = sum((s.predicted_prob - o) ** 2 for s, o in zip(synthetic, outcomes)) / len(synthetic)
    da = sum(1 for s in synthetic if s.direction_correct) / len(synthetic)
    return BaselineMetricBlock(n=len(steps), brier_score=bs, directional_accuracy=da)


def compute_naive_baselines(steps: list[BacktestStep]) -> NaiveBaselineReport:
    """Compute both coin-flip and last-period baselines in one call."""
    return NaiveBaselineReport(
        coin_flip=compute_coin_flip_baseline(steps),
        last_period=compute_last_period_baseline(steps),
    )


def naive_baseline_guard(
    arm_steps: list[BacktestStep],
    slack: float = 0.02,
) -> dict:
    """Strict quality gate: arm directional accuracy must beat baselines.

    Returns {passes, arm_dir_acc, baseline_dir_acc, gap}.
    The arm must satisfy: arm_dir_acc >= baseline_dir_acc - slack.
    """
    from research.backtest.metrics import directional_accuracy

    arm_dir_acc = directional_accuracy(arm_steps)
    baselines = compute_naive_baselines(arm_steps)
    baseline_dir_acc = max(
        baselines.coin_flip.directional_accuracy,
        baselines.last_period.directional_accuracy,
    )
    gap = arm_dir_acc - baseline_dir_acc
    return {
        "passes": gap >= -slack,
        "arm_dir_acc": arm_dir_acc,
        "baseline_dir_acc": baseline_dir_acc,
        "gap": gap,
    }
